moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1186 @@
+          '"' -> {
+            buffer.readByte() // consume the '\"'.
+            PEEKED_DOUBLE_QUOTED_NAME
+          }
+          '\'' -> {
+            buffer.readByte() // consume the '\''.
+            checkLenient()
+            PEEKED_SINGLE_QUOTED_NAME
+          }
+          '}' -> if (peekStack != JsonScope.NONEMPTY_OBJECT) {
+            buffer.readByte() // consume the '}'.
+            PEEKED_END_OBJECT
+          } else {
+            throw syntaxError("Expected name")
+          }
+          else -> {
+            checkLenient()
+            if (isLiteral(c.code)) {
+              PEEKED_UNQUOTED_NAME
+            } else {
+              throw syntaxError("Expected name")
+            }
+          }
+        }
+        peeked = next
+        return next
+      }
+      JsonScope.DANGLING_NAME -> {
+        scopes[stackSize - 1] = JsonScope.NONEMPTY_OBJECT
+        // Look for a colon before the value.
+        val c = nextNonWhitespace(true).toChar()
+        buffer.readByte() // Consume ':'.
+        when (c) {
+          ':' -> Unit /*no op*/
+          '=' -> {
+            checkLenient()
+            if (source.request(1) && buffer[0].asChar() == '>') {
+              buffer.readByte() // Consume '>'.
+            }
+          }
+          else -> throw syntaxError("Expected ':'")
+        }
+      }
+      JsonScope.EMPTY_DOCUMENT -> scopes[stackSize - 1] = JsonScope.NONEMPTY_DOCUMENT
+      JsonScope.NONEMPTY_DOCUMENT -> {
+        if (nextNonWhitespace(false) == -1) {
+          peeked = PEEKED_EOF
+          return peeked
+        } else {
+          checkLenient()
+        }
+      }
+      JsonScope.STREAMING_VALUE -> {
+        valueSource!!.discard()
+        valueSource = null
+        stackSize--
+        return doPeek()
+      }
+      else -> check(peekStack != JsonScope.CLOSED) { "JsonReader is closed" }
+    }
+    // "fallthrough" from previous `when`
+    when (nextNonWhitespace(true).toChar()) {
+      ']' -> {
+        if (peekStack == JsonScope.EMPTY_ARRAY) {
+          buffer.readByte() // Consume ']'.
+          peeked = PEEKED_END_ARRAY
+          return peeked
+        }
+        // In lenient mode, a 0-length literal in an array means 'null'.
+        return when (peekStack) {
+          JsonScope.EMPTY_ARRAY, JsonScope.NONEMPTY_ARRAY -> {
+            checkLenient()
+            peeked = PEEKED_NULL
+            peeked
+          }
+          else -> throw syntaxError("Unexpected value")
+        }
+      }
+      // In lenient mode, a 0-length literal in an array means 'null'.
+      ';', ',' -> return when (peekStack) {
+        JsonScope.EMPTY_ARRAY, JsonScope.NONEMPTY_ARRAY -> {
+          checkLenient()
+          peeked = PEEKED_NULL
+          peeked
+        }
+        else -> throw syntaxError("Unexpected value")
+      }
+      '\'' -> {
+        checkLenient()
+        buffer.readByte() // Consume '\''.
+        peeked = PEEKED_SINGLE_QUOTED
+        return peeked
```

Comment:
I know we changed this pattern from `return PEEKED_WHATEVER.also { peeked = it }`, but the new format (which I prefer visually) may be slower because it re-reads the value from the field rather than reusing the local. It may actually be worth switching back since performance is of the utmost importance at this layer.

@swankjesse opinions?

Reply:
is there a way to benchmark this? I don't think Moshi has any benchmarking code (besides #773)? 

Obviously I don't mind reverting this change -- mostly asking out of curiosity